### ISSUE TITLE
With threads, Switch.pm leaks scalars

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6014]
Migrated from [rt.perl.org#17964](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=17964) (status was 'resolved')

Searchable as RT17964$

**Comment by p5pRT** [comment]
### From madas@comcast.net

#### Created by madas@comcast.net



The following program generates a "Scalars leaked&#8203;: 1" error message\.

\-\-\-\-
\#\!/usr/bin/perl
use threads;
use Switch;
threads\->new\(sub\{\}\)\->join;
\-\-\-\-

This program does nothing but spawn a thread which immediately
terminates\.
Even though it doesn't actually use a switch statement\, just having a 
"use Switch;" triggers the "Scalars leaked&#8203;: 1" message\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.0:

Configured by Debian Project at Sat Sep 14 18:17:32 UTC 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i386-linux-thread-multi
    uname='linux cyberhq 2.4.19 #1 smp sun aug 4 11:30:45 pdt 2002 i686
unknown unknown gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN
-Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr
-Dprivlib=/usr/share/perl/5.8.0 -Darchlib=/usr/lib/perl/5.8.0
-Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5
-Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local
-Dsitelib=/usr/local/share/perl/5.8.0
-Dsitearch=/usr/local/lib/perl/5.8.0 -Dman1dir=/usr/share/man/man1
-Dman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3perl
-Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm
-Duseshrplib -Dlibperl=libperl.so.5.8.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN
-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing
-I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011002 (Debian prerelease)',
gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=true,
libperl=libperl.so.5.8.0
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /etc/perl
    /usr/local/lib/perl/5.8.0
    /usr/local/share/perl/5.8.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8.0
    /usr/share/perl/5.8.0
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/mds
    LANG=C
    LANGUAGE (unset)
    LC_CTYPE=
   
LD_LIBRARY_PATH=/home/mds/Work/SEL/libtg/lib:/home/mds/Work/SEL/libpith/lib
    LOGDIR (unset)
    PATH=~/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash


-- 
Marc Shapiro      | marc@sensics.com      | c: 410-499-0687
9 Cormer Ct #302  | madas@comcast.net     | h: 410-308-0199
Timonium MD 21093 | mds@lions.med.jhu.edu | http://www.baltolug.org/


```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 10&#8203;:11 AM 10/16/02 \+0000\, via RT wrote&#8203;:
>This program does nothing but spawn a thread which immediately
>terminates\.
>Even though it doesn't actually use a switch statement\, just having a
>"use Switch;" triggers the "Scalars leaked&#8203;: 1" message\.

Unfortunately\, there are many ways to create the Scalars leaked&#8203;: message if 
you're using threads\.  One of them is when you're doing funky stuff in 
BEGIN \(such as starting a thread\)\.  This may be related\, as Switch is a 
source\-filter \(right?\)\.


Liz

**Comment by p5pRT** [comment]
### From arthur@contiller.se


On onsdag\, okt 16\, 2002\, at 21&#8203;:52 Europe/Stockholm\, Elizabeth Mattijsen 
wrote&#8203;:

> Unfortunately\, there are many ways to create the Scalars leaked&#8203;: 
> message if you're using threads\.  One of them is when you're doing 
> funky stuff in BEGIN \(such as starting a thread\)\.  This may be 
> related\, as Switch is a source\-filter \(right?\)\.
>

Starting a thread from BEGIN is definitly not supported :\-\)

Arthur

**Comment by p5pRT** [comment]
### From @lizmat

At 10&#8203;:03 PM 10/16/02 \+0200\, Arthur Bergman wrote&#8203;:
>>Unfortunately\, there are many ways to create the Scalars leaked&#8203;: message 
>>if you're using threads\.  One of them is when you're doing funky stuff in 
>>BEGIN \(such as starting a thread\)\.  This may be related\, as Switch is a 
>>source\-filter \(right?\)\.
>Starting a thread from BEGIN is definitly not supported :\-\)

Seems to work fine with Thread&#8203;::Tie though\.\.\.


Liz

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Fixed by patch \#19301 that increases the reference count on the wierd rsfp entries\.

The test case that was needed for this was slightly different\, you had to add a \_\_DATA\_\_ 
or \_\_END\_\_ to it or do it from the console\.

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
